executor/builder.go

```diff
@@ -523,12 +523,29 @@ func (b *executorBuilder) buildRecoverIndex(v *plannercore.RecoverIndex) Executo
 		b.err = errors.Errorf("secondary index `%v` is not found in table `%v`", v.IndexName, v.Table.Name.O)
 		return nil
 	}
+	var containsGenedCol bool
+	cols := buildIdxColsConcatHandleCols(tblInfo, index.Meta(), func(indexInfo *model.IndexInfo) []*model.ColumnInfo {
+		columns := make([]*model.ColumnInfo, 0, len(indexInfo.Columns)+8)
+		for _, iCol := range indexInfo.Columns {
+			if tblInfo.Columns[iCol.Offset].IsGenerated() {
+				containsGenedCol = true
```

Comment:
Passing a function to `buildIdxColsConcatHandleCols` decrease the readability and make its name confusing.

I suggest extracting the initialization of `containsGenedCol` to a separate function, renaming `buildIdxColsConcatHandleCols` into `buildColsConcatHandleCols`, and passing a `hasVirtualCol bool` into it.

Reply:
ok

Reply:
done, @tangenta PTAL